src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,314 @@
+            _collectThread = new Thread(() => CollectWorker(_cts.Token));
+            _collectThread.Start();
+
+            _listener.Start();
+            _initialInstrumentEnumerationComplete();
+        }
+
+        private void CollectWorker(CancellationToken cancelToken)
+        {
+            try
+            {
+                double collectionIntervalSecs = -1;
+                lock (this)
+                {
+                    collectionIntervalSecs = _collectionPeriod.TotalSeconds;
+                }
+                if (collectionIntervalSecs < MinCollectionTimeSecs)
+                {
+                    // this is purely for defensive programming and it should be impossible
+                    // for a developer to observe this exception unless this library has a
+                    // bug in it. Because of that we didn't bother to localize the error message.
+                    throw new InvalidOperationException($"_collectionPeriod must be >= {MinCollectionTimeSecs} sec");
+                }
+
+                DateTime startTime = DateTime.UtcNow;
+                DateTime intervalStartTime = startTime;
+                while (!cancelToken.IsCancellationRequested)
+                {
+                    // intervals end at startTime + X*collectionIntervalSecs. Under normal
+                    // circumstance X increases by 1 each interval, but if the time it
+                    // takes to do collection is very large then we might need to skip
+                    // ahead multiple intervals to catch back up.
+                    //
+                    DateTime now = DateTime.UtcNow;
+                    double secsSinceStart = (now - startTime).TotalSeconds;
+                    double alignUpSecsSinceStart = Math.Ceiling(secsSinceStart / collectionIntervalSecs) *
+                        collectionIntervalSecs;
+                    DateTime nextIntervalStartTime = startTime.AddSeconds(alignUpSecsSinceStart);
+
+                    // The delay timer precision isn't exact. We might have a situation
+                    // where in the previous loop iterations intervalStartTime=20.00,
+                    // nextIntervalStartTime=21.00, the timer was supposed to delay for 1s but
+                    // it exited early so we looped around and DateTime.Now=20.99.
+                    // Aligning up from DateTime.Now would give us 21.00 again so we also need to skip
+                    // forward one time interval
+                    DateTime minNextInterval = intervalStartTime.AddSeconds(collectionIntervalSecs);
+                    if (nextIntervalStartTime <= minNextInterval)
+                    {
+                        nextIntervalStartTime = minNextInterval;
+                    }
+
+                    // pause until the interval is complete
+                    TimeSpan delayTime = nextIntervalStartTime - now;
+                    cancelToken.WaitHandle.WaitOne(delayTime);
```

Comment:
I think this may be the only use of CancellationToken.WaitHandle in all of our product source. :-)

Reply:
I was hoping for Thread.Sleep(timeout, cancelToken) originally, but this also did what I needed it to do.